lib/middleware/remote_user_auth_middleware.rb

```diff
@@ -0,0 +1,53 @@
+require 'securerandom'
+
+class RemoteUserAuthMiddleware
+
+  DefaultUserHeaderName = "Remote-User"
+  DefaultEmailHeaderName = "Remote-Email"
+  DefaultGroupsHeaderName = "Remote-Groups"
+  # if the header specified by user_header_name is provided in the
+  # request, treat the value of that header as a username, and log in as that
+  # user, creating if necessary.
+  # 
+  # This middleware can also be configured to detect headers containing the
+  # user's email, and comma separated group membership (along with a group name
+  # that signifies the created user should be an admin) during the user
+  # creation phase.
+  #
+  def initialize(app, user_header_name = nil, email_header_name = nil, groups_header_name = nil, admin_group = nil, warden_scope = :user)
+    @app = app
+    @user_header_name = (user_header_name || DefaultUserHeaderName).gsub("-", "_").upcase
+    @email_header_name = (email_header_name || DefaultEmailHeaderName).gsub("-", "_").upcase
+    @groups_header_name = (groups_header_name || DefaultGroupsHeaderName).gsub("-", "_").upcase
+    @admin_group = admin_group
+    @warden_scope = warden_scope
+  end
+
+  def call(env)
+    begin
+      username = env['HTTP_' + @user_header_name]
+      # do nothing if no remote-user is presented
+      if username
+        user = User.find_or_create_by!(username: username) do |u|
+          u.requires_no_invitation_code!
+
+          # needed for devise validations
+          u.password = SecureRandom.urlsafe_base64(24 + rand(8))
+          u.email = env['HTTP_' + @email_header_name] || "#{username}@remoteuser.auth"
+          u.admin = !!@admin_group && env['HTTP_' + @groups_header_name].split(",").include?(@admin_group)
```

Comment:
This assumes that the header has no spaces after the comma: `group1, group2` will not find `group2`. This falls back to a secure default ("not admin"), but still may be unexpected

Reply:
this functionality needs to be built basically case-by-case, and IMO it's a fairly edge use case. My auth proxy returns groups as comma-separated, and while you are right that other auth proxies may return comma-space separated lists, yet others may return pipe-separated, slash-separated, plus-separated, or any other thing. 

If you have a specific use case for comma-space separated, I'm happy to add it, but meanwhile I don't think it's useful to try to predict out-of-hand what the delimiter would be. I'll separate it out to make future modifications easier.

Reply:
I just stumbled over it, no preference in any way

Reply:
oh lol I thought you were a maintainer! yeah this doesn't look like it'll ever get merged. I opened in summer 2022. I don't even use this software any more, I just pushed edits to do right by my original effort in putting the PR together.

Reply:
Haha! No, I was searching  if there was support for SSO and then landed here. „Drive-by review“, so to say